{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job started at:\n",
      "16:08:41.175603\n",
      "\n",
      "This is the raw_dir:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary/\n",
      "\n",
      "This is the record_dir:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//record/\n",
      "\n",
      "The log_dir is:\n",
      "/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/\n",
      "\n",
      "The script_dir is:\n",
      "/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1/\n",
      "\n",
      "The fastq_dir is:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq/\n",
      "\n",
      "The ribo_dir is:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1/\n",
      "\n",
      "The gc_dir is:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1/\n",
      "\n",
      "The htseq_dir is:\n",
      "/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//htseq/exp9_test1/\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT8.bam\n",
      "\n",
      "he u_id is: prPT8\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT8.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT8/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT8/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT8.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.2.fastq.gz']\n",
      "\n",
      "Submitting job 0 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT8.bam']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.2.fastq.gz']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j0_prPT8  -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 2  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash 2 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT8.bam /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.unpaired.2.fastq.gz\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT8/']\n",
      "\n",
      "At least one job 1 input file does not exist, holding job\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT8.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT8/']\n",
      "\n",
      "At least one job 2 input file does not exist, holding job\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//FT3.bam\n",
      "\n",
      "he u_id is: FT3\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//FT3.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//FT3/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//FT3/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//FT3.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.unpaired.2.fastq.gz']\n",
      "\n",
      "Job 0 has previously been completed, no need to run\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//FT3/']\n",
      "\n",
      "Submitting job 1 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//FT3/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j1_FT3 -hold_jid job0_FT3 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 8  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash 8 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//FT3/\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//FT3/']\n",
      "\n",
      "Submitting job 2 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//FT3/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j2_FT3 -hold_jid job1_FT3 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 6  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash 6 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//FT3.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//FT3/\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT1.bam\n",
      "\n",
      "he u_id is: rfPT1\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT1.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT1/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT1/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT1.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.unpaired.2.fastq.gz']\n",
      "\n",
      "Job 0 has previously been completed, no need to run\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT1/']\n",
      "\n",
      "Submitting job 1 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT1/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j1_rfPT1 -hold_jid job0_rfPT1 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 8  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash 8 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT1/\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT1/']\n",
      "\n",
      "Submitting job 2 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT1/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j2_rfPT1 -hold_jid job1_rfPT1 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 6  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash 6 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT1.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT1/\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT3.bam\n",
      "\n",
      "he u_id is: rfPT3\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT3.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT3/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT3/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//rfPT3.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.unpaired.2.fastq.gz']\n",
      "\n",
      "Job 0 has previously been completed, no need to run\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT3/']\n",
      "\n",
      "Submitting job 1 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT3/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j1_rfPT3 -hold_jid job0_rfPT3 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 8  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash 8 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//rfPT3/\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT3/']\n",
      "\n",
      "Submitting job 2 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT3/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j2_rfPT3 -hold_jid job1_rfPT3 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 6  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash 6 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//rfPT3.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//rfPT3/\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT11.bam\n",
      "\n",
      "he u_id is: prPT11\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT11.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT11/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT11/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT11.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.unpaired.2.fastq.gz']\n",
      "\n",
      "Job 0 has previously been completed, no need to run\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT11/']\n",
      "\n",
      "Submitting job 1 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT11/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j1_prPT11 -hold_jid job0_prPT11 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 8  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash 8 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT11/\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT11/']\n",
      "\n",
      "Submitting job 2 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT11/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j2_prPT11 -hold_jid job1_prPT11 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 6  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash 6 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT11.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT11/\n",
      "\n",
      "\n",
      "The infile is /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary/prPT9.bam\n",
      "\n",
      "he u_id is: prPT9\n",
      "\n",
      "cores:\n",
      "['2', '8', '6']\n",
      "\n",
      "inputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT9.bam', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz']\n",
      "\n",
      "inputs2:\n",
      "['none', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz']\n",
      "\n",
      "inputs3:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "input4:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "outputs1:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT9/', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT9/']\n",
      "\n",
      "outputs2:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs3:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.unpaired.1.fastq.gz', 'none', 'none']\n",
      "\n",
      "outputs4:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.unpaired.2.fastq.gz', 'none', 'none']\n",
      "\n",
      "scripts:\n",
      "['/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//0.bamtofastq.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash', '/Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash']\n",
      "\n",
      "qsub_params:\n",
      "['none', 'none', 'none']\n",
      "\n",
      "\n",
      "For job number 0:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//prPT9.bam']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.unpaired.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.unpaired.2.fastq.gz']\n",
      "\n",
      "Job 0 has previously been completed, no need to run\n",
      "\n",
      "\n",
      "For job number 1:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT9/']\n",
      "\n",
      "Submitting job 1 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT9/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j1_prPT9 -hold_jid job0_prPT9 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 8  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//1.starRibo.bash 8 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/ribo/exp9_test1//prPT9/\n",
      "\n",
      "\n",
      "For job number 2:\n",
      "\n",
      "input/s or in dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz']\n",
      "\n",
      "output/s or out dir/s are: ['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT9/']\n",
      "\n",
      "Submitting job 2 with input/s or in dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz', '/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz']\n",
      "\n",
      "and output/s or out dir/s:\n",
      "['/Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT9/']\n",
      "\n",
      "/opt/gridengine/bin/linux-x64/qsub -q long.q -N j2_prPT9 -hold_jid job1_prPT9 -b y -wd /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//logs/ -j y -R y -pe smp 6  -V /Users/jamestorpy/clusterHome2//projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/exp9_test1//2.starGC.bash 6 /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.1.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//raw_files/fullsamples/bowtell_primary//fastq//prPT9.2.fastq.gz /Users/jamestorpy/clusterHome//projects/hgsoc_repeats/RNA-seq//results//star/GC/exp9_test1//prPT9/\n"
     ]
    }
   ],
   "source": [
    "#!/share/ClusterShare/software/contrib/marsmi/python/3.5.1/python\n",
    "\n",
    "### 1.bamtofastq_autosubmit.py ###\n",
    "\n",
    "# This script detects which output files still must be created from the repeatsPipeline and submits jobs for them #\n",
    "\n",
    "### 0. Define starting variables ###\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import re\n",
    "\n",
    "# set up bash environment:\n",
    "os.system('source /home/jamtor/.bashrc')\n",
    "os.system('echo $SGE_ROOT')\n",
    "os.system('module load gi/zlib/1.2.8')\n",
    "os.system('module load phuluu/samtools/1.4')\n",
    "\n",
    "print('Job started at:')\n",
    "print(datetime.datetime.now().time())\n",
    "\n",
    "# make directory hierachy:\n",
    "projectname = 'hgsoc_repeats'\n",
    "sample_type = 'fullsamples/bowtell_primary'\n",
    "R_dir = '/home/jamtor/local/lib/r/R-3.2.2/bin/R/'\n",
    "exp_name = 'exp9_test1'\n",
    "total_no_jobs = 3\n",
    "q = 'long'\n",
    "\n",
    "# home_dir is where input/intermediate files are located:\n",
    "#home_dir = '/share/ScratchGeneral/jamtor/'\n",
    "home_dir = '/Users/jamestorpy/clusterHome/'\n",
    "# home_dir2 is where scripts are located and final outputs will be stored:\n",
    "#home_dir2 = '/share/ClusterShare/thingamajigs/jamtor'\n",
    "home_dir2 = '/Users/jamestorpy/clusterHome2/'\n",
    "\n",
    "project_dir = home_dir + '/projects/' + projectname + '/RNA-seq/'\n",
    "raw_dir = project_dir + '/raw_files/' + sample_type + '/'\n",
    "results_dir = project_dir + '/results/'\n",
    "record_dir = project_dir + '/record/'\n",
    "\n",
    "os.makedirs(record_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('This is the raw_dir:')\n",
    "print(raw_dir)\n",
    "\n",
    "print('')\n",
    "print('This is the record_dir:')\n",
    "print(record_dir)\n",
    "\n",
    "# scripts/logs directories:\n",
    "script_dir= home_dir2 + '/projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/' + exp_name + '/'\n",
    "log_dir=script_dir + '/logs/'\n",
    "os.makedirs(log_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('The log_dir is:')\n",
    "print(log_dir)\n",
    "print('')\n",
    "print('The script_dir is:')\n",
    "print(script_dir)\n",
    "\n",
    "# define other output directories:\n",
    "fastq_dir=raw_dir + '/fastq/'\n",
    "os.makedirs(fastq_dir, exist_ok=True)\n",
    "\n",
    "ribo_dir = results_dir + '/star/ribo/' + exp_name + '/'\n",
    "os.makedirs(ribo_dir, exist_ok=True)\n",
    "\n",
    "gc_dir = results_dir + '/star/GC/' + exp_name + '/'\n",
    "os.makedirs(gc_dir, exist_ok=True)\n",
    "\n",
    "htseq_dir = results_dir + '/htseq/' + exp_name + '/'\n",
    "os.makedirs(htseq_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('The fastq_dir is:')\n",
    "print(fastq_dir)\n",
    "print('')\n",
    "print('The ribo_dir is:')\n",
    "print(ribo_dir)\n",
    "print('')\n",
    "print('The gc_dir is:')\n",
    "print(gc_dir)\n",
    "print('')\n",
    "print('The htseq_dir is:')\n",
    "print(htseq_dir)\n",
    "\n",
    "### 2. Define specific job parameters for each sample ###\n",
    "\n",
    "# read in raw files and fetch ids:\n",
    "#in_files = glob.glob(raw_dir + '*.bam')\n",
    "in_files = [raw_dir + '/prPT8.bam', raw_dir + '/FT3.bam', raw_dir + '/rfPT1.bam',     raw_dir + '/rfPT3.bam', raw_dir + '/prPT11.bam', raw_dir + 'prPT9.bam']\n",
    "\n",
    "for infile in in_files:\n",
    "    print('')\n",
    "    print('')\n",
    "    print('The infile is ' + infile)\n",
    "    u_id = re.sub(\n",
    "        '.bam', '', os.path.basename(infile)\n",
    "    )\n",
    "    print('')\n",
    "    print('he u_id is: ' + u_id)\n",
    "\n",
    "    \n",
    "    ### job 0 inputs ###\n",
    "    # fastq to bam conversion #\n",
    "    \n",
    "    cores = ['2']\n",
    "\n",
    "    inputs1 = [raw_dir + '/' + u_id + '.bam']\n",
    "    inputs2 = ['none']\n",
    "    inputs3 = ['none']\n",
    "    inputs4 = ['none']\n",
    "\n",
    "    outputs1 = [fastq_dir + '/' + u_id + '.1.fastq.gz']\n",
    "    outputs2 = [fastq_dir + '/' + u_id + '.2.fastq.gz']\n",
    "    outputs3 = [fastq_dir + '/' + u_id + '.unpaired.1.fastq.gz']\n",
    "    outputs4 = [fastq_dir + '/' + u_id + '.unpaired.2.fastq.gz']\n",
    "    \n",
    "    scripts = [script_dir + '/0.bamtofastq.bash']\n",
    "    \n",
    "    qsub_params = ['none']\n",
    "    \n",
    "    \n",
    "    ## job 1 inputs ##\n",
    "    # alignment of fastqs to ribosomal reference #\n",
    "    \n",
    "    cores.append('8')\n",
    "    \n",
    "    inputs1.append(fastq_dir + '/' + u_id + '.1.fastq.gz')\n",
    "    inputs2.append(fastq_dir + '/' + u_id + '.2.fastq.gz')\n",
    "    inputs3.append('none')\n",
    "    inputs4.append('none')\n",
    "    \n",
    "    outputs1.append(ribo_dir + '/' + u_id + '/')\n",
    "    outputs2.append('none')\n",
    "    outputs3.append('none')\n",
    "    outputs4.append('none')\n",
    "    \n",
    "    scripts.append(script_dir + '/1.starRibo.bash')\n",
    "    \n",
    "    qsub_params.append('none')\n",
    "    \n",
    "    \n",
    "    ## job 2 inputs ##\n",
    "    # alignment of fastqs to gc reference #\n",
    "        \n",
    "    cores.append('6')\n",
    "    \n",
    "    inputs1.append(fastq_dir + '/' + u_id + '.1.fastq.gz')\n",
    "    inputs2.append(fastq_dir + '/' + u_id + '.2.fastq.gz')\n",
    "    inputs3.append('none')\n",
    "    inputs4.append('none')\n",
    "    \n",
    "    outputs1.append(gc_dir + '/' + u_id + '/')\n",
    "    outputs2.append('none')\n",
    "    outputs3.append('none')\n",
    "    outputs4.append('none')\n",
    "    \n",
    "    scripts.append(script_dir + '/2.starGC.bash')\n",
    "    \n",
    "    qsub_params.append('none')\n",
    "    \n",
    "    \n",
    "    ## job 3 inputs ##\n",
    "    # alignment of fastqs to gc reference #\n",
    "        \n",
    "    cores.append('7')\n",
    "    \n",
    "    inputs1.append(gc_dir + '/' + u_id + '/Aligned.out.bam')\n",
    "    inputs2.append('none')\n",
    "    inputs3.append('none')\n",
    "    inputs4.append('none')\n",
    "    \n",
    "    outputs1.append(gc_dir + '/' + u_id + '/Aligned.novosorted.by.name.out.bam')\n",
    "    outputs2.append('none')\n",
    "    outputs3.append('none')\n",
    "    outputs4.append('none')\n",
    "    \n",
    "    scripts.append(script_dir + '/3.novosortName.bash')\n",
    "    \n",
    "    qsub_params.append('24G')\n",
    "    \n",
    "    \n",
    "    print('')\n",
    "    print('cores:')\n",
    "    print(cores)\n",
    "    print('')\n",
    "    print('inputs1:')\n",
    "    print(inputs1)\n",
    "    print('')\n",
    "    print('inputs2:')\n",
    "    print(inputs2)\n",
    "    print('')\n",
    "    print('inputs3:')\n",
    "    print(inputs3)\n",
    "    print('')\n",
    "    print('input4:')\n",
    "    print(inputs4)\n",
    "    print('')\n",
    "    print('outputs1:')\n",
    "    print(outputs1)\n",
    "    print('')\n",
    "    print('outputs2:')\n",
    "    print(outputs2)\n",
    "    print('')\n",
    "    print('outputs3:')\n",
    "    print(outputs3)\n",
    "    print('')\n",
    "    print('outputs4:')\n",
    "    print(outputs4)\n",
    "    print('')\n",
    "    print('scripts:')\n",
    "    print(scripts)\n",
    "    print('')\n",
    "    print('qsub_params:')\n",
    "    print(qsub_params)\n",
    "    \n",
    "    for i in range(total_no_jobs):\n",
    "        print('')\n",
    "        print('')\n",
    "        print('For job number ' + str(i) + ':')\n",
    "        \n",
    "        # set up job holding if necessary:\n",
    "        if i==0:\n",
    "            holdj = ' '\n",
    "        else:\n",
    "            holdj = ' -hold_jid job' + str(i-1) + '_' + u_id\n",
    "            \n",
    "        # define qsub_params:\n",
    "        if qsub_params:\n",
    "            if qsub_params[i] == 'none':\n",
    "                qsub_params = ''\n",
    "        \n",
    "        inputs = [inputs1[i]]\n",
    "        for input in [inputs2[i], inputs3[i], inputs4[i]]:\n",
    "            if input !='none':\n",
    "                inputs.append(input)\n",
    "        print('')\n",
    "        print('input/s or in dir/s are: ' + str(inputs))\n",
    "        \n",
    "        outputs = [outputs1[i]]\n",
    "        for output in [outputs2[i], outputs3[i], outputs4[i]]:\n",
    "            if output !='none':\n",
    "                outputs.append(output)\n",
    "        print('')\n",
    "        print('output/s or out dir/s are: ' + str(outputs))\n",
    "    \n",
    "        print('')\n",
    "        if not all([os.path.isfile(inp) for inp in inputs]):\n",
    "            print('At least one job ' + str(i) + ' input file does not exist, holding job')\n",
    "        \n",
    "        elif all([os.path.isfile(inp) for inp in inputs]) and not all([os.path.isfile(outp)             for outp in outputs]):\n",
    "            print('Submitting job ' + str(i) + ' with input/s or in dir/s:')\n",
    "            print(inputs)\n",
    "            print('')\n",
    "            print('and output/s or out dir/s:')\n",
    "            print(outputs)\n",
    "            \n",
    "            print('')\n",
    "            print('/opt/gridengine/bin/linux-x64/qsub '                   + '-q ' + q + '.q '                   + '-N j' + str(i) + '_' + u_id                   + holdj                   + ' -b y -wd '                   + log_dir                   + ' -j y -R y '\n",
    "                  + '-pe smp ' + cores[i] + ' ' \\\n",
    "                  + qsub_params \\\n",
    "                  + ' -V ' + scripts[i] + ' ' + cores[i] + ' ' + ' '.join(inputs) + ' ' + ' '.join(outputs))\n",
    "            \n",
    "        elif all([os.path.isfile(outp) for outp in outputs]):\n",
    "            print('Job ' + str(i) + ' has previously been completed, no need to run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook job_autosubmit.ipynb to script\n",
      "[NbConvertApp] Writing 22401 bytes to job_autosubmit.py\n"
     ]
    }
   ],
   "source": [
    "!jupyter nbconvert --to script job_autosubmit.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    " ### 2. Submit jobs ###\n",
    "    \n",
    "    for i in range(total_no_jobs):\n",
    "\n",
    "                         \n",
    "        if not os.path.isfile(job_input):\n",
    "            print('Job ' i ' input file ' + job_input + 'does not exist, holding job')\n",
    "        elif:\n",
    "            if not os.path.isfile(job_output1) and os.path.isfile(job_input):\n",
    "            print('Submitting job ' + i 'with input ' + job_input ' and outputs ' + \\\n",
    "                  job_output1 + ' ' + job_output2 + ' ' + job_output3 + ' ' + job_output4)\n",
    "            os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + q + '.q -N job' + i + '_' \\\n",
    "                      + u_id + ' -b y -wd ' + log_dir + ' -j y -R y -pe smp ' + job_cores + \\\n",
    "                      ' -V \n",
    "                      \n",
    "                      /home/jamtor/local/bin/bamtofastq F=/share/ScratchGeneral/jamtor/projects/hgsoc_repeats/RNA-seq/raw_files/fullsamples/bowtell_primary/fastq/FT3.1.fastq.gz F2=/share/ScratchGeneral/jamtor/projects/hgsoc_repeats/RNA-seq/raw_files/fullsamples/bowtell_primary/fastq/FT3.2.fastq.gz O=/share/ScratchGeneral/jamtor/projects/hgsoc_repeats/RNA-seq/raw_files/fullsamples/bowtell_primary/fastq/FT3.unpaired.1.fastq.gz O2=/share/ScratchGeneral/jamtor/projects/hgsoc_repeats/RNA-seq/raw_files/fullsamples/bowtell_primary/fastq/FT3.unpaired.2.fastq.gz filename=/share/ScratchGeneral/jamtor/projects/hgsoc_repeats/RNA-seq/raw_files/fullsamples/bowtell_primary/FT3.bam gz=1)\n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-1-1a29390cb688>, line 159)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-1a29390cb688>\"\u001b[0;36m, line \u001b[0;32m159\u001b[0m\n\u001b[0;31m    os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N fq_' + u_id + ' -b y -wd ' + log_dir +                           '-j y -R y -pe smp ' + fqcores + ' -V bamtofastq F=' +                           fastq_dir + '/' + u_id + '.1.fastq.gz F2=' + fastq_dir + '/' +                           u_id + '.2.fastq.gz O= + 'fastq_dir + '/' + u_id +                           '.unpaired.1.fastq.gz O2=' + fastq_dir + '/' +                           u_id + '.unpaired.2.fastq.gz filename=' + raw_dir + '/' + u_id +\u001b[0m\n\u001b[0m                                                                                                                                                                                                                                                                                                                                                        ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#!/share/ClusterShare/software/contrib/marsmi/python/3.5.1/python\n",
    "\n",
    "### 1.bamtofastq_autosubmit.py ###\n",
    "\n",
    "# This script detects which output files still must be created from the repeatsPipeline and submits jobs for them #\n",
    "\n",
    "### 0. Define starting variables ###\n",
    "\n",
    "import os\n",
    "import datetime\n",
    "from path import path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# set up bash environment:\n",
    "os.system('source /home/jamtor/.bashrc')\n",
    "os.system('echo $SGE_ROOT')\n",
    "os.system('module load gi/zlib/1.2.8')\n",
    "os.system('module load phuluu/samtools/1.4')\n",
    "\n",
    "print('Job started at:')\n",
    "print(datetime.datetime.now().time())\n",
    "\n",
    "# make directory hierachy:\n",
    "projectname = 'hgsoc_repeats'\n",
    "sample_type = 'fullsamples/bowtell_primary'\n",
    "R_dir = '/home/jamtor/local/lib/r/R-3.2.2/bin/R/'\n",
    "\n",
    "# home_dir is where input/intermediate files are located:\n",
    "#home_dir = '/share/ScratchGeneral/jamtor/'\n",
    "home_dir = '/Users/jamestorpy/clusterHome/'\n",
    "# home_dir2 is where scripts are located and final outputs will be stored:\n",
    "#home_dir2 = '/share/ClusterShare/thingamajigs/jamtor'\n",
    "home_dir2 = '/Users/jamestorpy/clusterHome2/'\n",
    "\n",
    "project_dir = home_dir + '/projects/' + projectname + '/RNA-seq/'\n",
    "raw_dir = project_dir + '/raw_files/' + sample_type + '/'\n",
    "results_dir = project_dir + '/results/'\n",
    "record_dir = project_dir + '/record/'\n",
    "\n",
    "os.makedirs(record_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('This is the raw_dir:')\n",
    "print(raw_dir)\n",
    "\n",
    "print('')\n",
    "print('This is the record_dir:')\n",
    "print(record_dir)\n",
    "\n",
    "# scripts/logs directories:\n",
    "scripts_dir= home_dir2 + '/projects/hgsoc_repeats/RNA-seq/scripts/repeatsPipeline/' + exp_name + '/'\n",
    "log_dir=scripts_dir + '/logs/'\n",
    "os.makedirs(log_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('The log_dir is:')\n",
    "print(log_dir)\n",
    "print('')\n",
    "print('The scripts_dir is:')\n",
    "print(scripts_dir)\n",
    "\n",
    "# define other output directories:\n",
    "fastq_dir=raw_dir + '/fastq/'\n",
    "os.makedirs(fastq_dir, exist_ok=True)\n",
    "\n",
    "ribo_dir = results_dir + '/star/ribo/' + exp_name + '/'\n",
    "os.makedirs(ribo_dir, exist_ok=True)\n",
    "\n",
    "gc_dir = results_dir + '/star/GC/' + exp_name + '/'\n",
    "os.makedirs(gc_dir, exist_ok=True)\n",
    "\n",
    "htseq_dir = results_dir + '/htseq/' + exp_name + '/'\n",
    "os.makedirs(htseq_dir, exist_ok=True)\n",
    "\n",
    "print('')\n",
    "print('The fastq_dir is:')\n",
    "print(fastq_dir)\n",
    "print('')\n",
    "print('The ribo_dir is:')\n",
    "print(ribo_dir)\n",
    "print('')\n",
    "print('The gc_dir is:')\n",
    "print(gc_dir)\n",
    "print('')\n",
    "print('The htseq_dir is:')\n",
    "print(htseq_dir)\n",
    "\n",
    "# specify cores needed:\n",
    "job1cores = '2'\n",
    "job2cores = '10'\n",
    "job3cores = '6'\n",
    "job4cores = '7'\n",
    "job5cores = '12'\n",
    "job6cores = '12'\n",
    "job7cores = '12'\n",
    "\n",
    "\n",
    "### 1. Define specific job parameters for each sample\n",
    "\n",
    "# read in raw files and fetch ids:\n",
    "os.list.dir()\n",
    "\n",
    "\n",
    "# specify job 1 input, output, command, and any additional qsub parameters:\n",
    "job1_in = fastq_dir/\n",
    "\n",
    "\n",
    "# specify any additional memory parameters:\n",
    "mem = '45G'\n",
    "sub_type = 'long'\n",
    "exp_name = 'exp9'\n",
    "draft='one'\n",
    "\n",
    "\n",
    "\n",
    "### 1. Load in record and job info files: ###\n",
    "\n",
    "# save entire qstat output:\n",
    "#os.system(\"qstat -u '*' | grep jamtor > \" + record_dir + \"/qstat.txt\")\n",
    "\n",
    "# save job ids of qstat:\n",
    "qstat = pd.read_csv(record_dir + '/qstat.txt', header = None)\n",
    "qstat_jid = list(qstat[0].str.split(' ').str[3])\n",
    "\n",
    "# read in completion record:\n",
    "comp_record = pd.read_table(record_dir + '/completion_record.tab')\n",
    "comp_record.set_index('samples', inplace = True)\n",
    "\n",
    "# save qstat cores output as file:\n",
    "#os.system(\"qstat -u '*' | grep jamtor | grep -e 'q@' | grep -v QRLOGIN | awk '{print $9}' > \" + script_dir + \"/qstat.txt\")\n",
    "\n",
    "# fetch core number:\n",
    "#slots = 0\n",
    "#with open(script_dir + '/qstat.txt') as f:\n",
    "#    for line in f:\n",
    "#        slots = slots + int(re.sub('\\\\n', '', line))\n",
    " \n",
    "slots = 80\n",
    "\n",
    "        \n",
    "### 2. If number of cores being used < 85, submit jobs that are not\n",
    "# complete and not currently queued/running ###\n",
    "\n",
    "if slots < 85:\n",
    "    for r in comp_record.iterrows():\n",
    "        u_id = r[0]\n",
    "        print('')\n",
    "        print('')\n",
    "        print('u_id is ' + u_id)\n",
    "        rep = r[1][0]\n",
    "        ribo = r[1][1]\n",
    "        gc = r[1][2]\n",
    "        all_htseq = r[1][3]\n",
    "        custom3_htseq = r[1][4]\n",
    "        gc_htseq = r[1][5]\n",
    "        \n",
    "        if ribo == False:\n",
    "            if 'fq_' + u_id in qstat_jid:\n",
    "                print('')\n",
    "                print('fq_' + u_id + ' already running')\n",
    "                os.system     \n",
    "            else:\n",
    "                print('')\n",
    "                print('Submitting fastq job')\n",
    "                os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N fq_' + u_id + ' -b y -wd ' + log_dir + \\\n",
    "                          '-j y -R y -pe smp ' + fqcores + ' -V bamtofastq F=' + \\\n",
    "                          fastq_dir + '/' + u_id + '.1.fastq.gz F2=' + fastq_dir + '/' + \\\n",
    "                          u_id + '.2.fastq.gz O= + 'fastq_dir + '/' + u_id + \\\n",
    "                          '.unpaired.1.fastq.gz O2=' + fastq_dir + '/' + \\\n",
    "                          u_id + '.unpaired.2.fastq.gz filename=' + raw_dir + '/' + u_id + \n",
    "                          '.bam gz=1')\n",
    "            if 'j3a_' + u_id in qstat_jid:\n",
    "                print('')\n",
    "                print('j3a_' + u_id + 'already running')\n",
    "            else:\n",
    "                print('')\n",
    "                print('Submitting star ribo job')\n",
    "                #os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N j3a_' + \\\n",
    "                #          u_id + ' -hold_jid fq_' + u_id + '-b y -wd ' log_dir + \\\n",
    "                #          ' -j y -R y -pe smp ' + ribocores + '-V ' + scripts_dir + \n",
    "                #          '/3a.starRibo.bash ' + u_id + ' ' + ribocores + ' ' + fastq_dir + ' ' + ribo_dir)\n",
    "            if gc == False:\n",
    "                if 'j3b_' + u_id in qstat_jid:\n",
    "                    print('')\n",
    "                    print('j3b_' + u_id + ' already running')\n",
    "                elif 'j3c_' + u_id in qstat_jid:\n",
    "                    print('')\n",
    "                    print('j3c_' + u_id + ' already running')\n",
    "                else:\n",
    "                    print('')\n",
    "                    print('Submitting star gc and sort jobs')\n",
    "                    #os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N j3b_' + \\\n",
    "                    #      u_id + ' -hold_jid fq_' + u_id + '-b y -wd ' log_dir + \\\n",
    "                    #      ' -j y -R y -pe smp ' + gccores + '-V ' + scripts_dir + \n",
    "                    #      '/3b.star_gc.py ' + u_id + ' ' + gccores + ' ' + fastq_dir)\n",
    "                    #os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N j3c_' + \\\n",
    "                    #      u_id + ' -hold_jid j3b_' + u_id + '-b y -wd ' log_dir + \\\n",
    "                    #      ' -j y -R y -pe smp ' + ribocores + '-V ' + scripts_dir + \n",
    "                    #      '/3c.novosort.py ' + u_id + ' ' + sortcores)\n",
    "                    \n",
    "                    if all_htseq == False or custom3_htseq == False or gc_htseq == False:\n",
    "                        if all_htseq == False:\n",
    "                            if 'j5a_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5a_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting all_htseq job')\n",
    "                        if custom3_htseq == False:\n",
    "                            if 'j5c_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5c_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting custom3_htseq job')\n",
    "                        if gc_htseq == False:\n",
    "                            if 'j5d_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5d_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting gc_htseq job')\n",
    "            elif gc == True:\n",
    "                if all_htseq == False or custom3_htseq == False or gc_htseq == False:\n",
    "                    if all_htseq == False:\n",
    "                        if 'j5a_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5a_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting all_htseq job')\n",
    "                    if custom3_htseq == False:\n",
    "                        if 'j5c_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5c_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting custom3_htseq job')\n",
    "                    if gc_htseq == False:\n",
    "                        if 'j5d_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5d_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting gc_htseq job')\n",
    "        elif ribo == True:\n",
    "            if gc == False:\n",
    "                if 'fq_' + u_id in qstat_jid:\n",
    "                    print('')\n",
    "                    print('fq_' + u_id + ' already running')\n",
    "                else:\n",
    "                    print('')\n",
    "                    print('Submitting fastq job')\n",
    "                    os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N fq_' + u_id + ' -b y -wd ' + log_dir + \\\n",
    "                          '-j y -R y -pe smp ' + fqcores + ' -V bamtofastq F=' + \\\n",
    "                          fastq_dir + '/' + u_id + '.1.fastq.gz F2=' + fastq_dir + '/' + \\\n",
    "                          u_id + '.2.fastq.gz O= + 'fastq_dir + '/' + u_id + \\\n",
    "                          '.unpaired.1.fastq.gz O2=' + fastq_dir + '/' + \\\n",
    "                          u_id + '.unpaired.2.fastq.gz filename=' + raw_dir + '/' + u_id + \n",
    "                          '.bam gz=1')\n",
    "                if 'j3b_' + u_id in qstat_jid:\n",
    "                    print('')\n",
    "                    print('j3b_' + u_id + ' already running')\n",
    "                elif 'j3c_' + u_id in qstat_jid:\n",
    "                    print('')\n",
    "                    print('j3c_' + u_id + ' already running')\n",
    "                else:\n",
    "                    print('')\n",
    "                    print('Submitting star gc and sort jobs')\n",
    "                    #os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N j3b_' + \\\n",
    "                    #      u_id + ' -hold_jid fq_' + u_id + '-b y -wd ' log_dir + \\\n",
    "                    #      ' -j y -R y -pe smp ' + gccores + '-V ' + scripts_dir + \n",
    "                    #      '/3b.star_gc.py ' + u_id + ' ' + gccores)\n",
    "                    #os.system('/opt/gridengine/bin/linux-x64/qsub -q ' + sub_type + '.q -N j3c_' + \\\n",
    "                    #      u_id + ' -hold_jid j3b_' + u_id + '-b y -wd ' log_dir + \\\n",
    "                    #      ' -j y -R y -pe smp ' + ribocores + '-V ' + scripts_dir + \n",
    "                    #      '/3c.novosort.py ' + u_id + ' ' + sortcores)\n",
    "                    if all_htseq == False or custom3_htseq == False or gc_htseq == False:\n",
    "                        if all_htseq == False:\n",
    "                            if 'j5a_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5a_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting all_htseq job')\n",
    "                        if custom3_htseq == False:\n",
    "                            if 'j5c_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5c_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting custom3_htseq job')\n",
    "                        if gc_htseq == False:\n",
    "                            if 'j5d_' + u_id in qstat_jid:\n",
    "                                print('')\n",
    "                                print('j5d_' + u_id + 'already running')\n",
    "                            else:\n",
    "                                print('')\n",
    "                                print('Submitting gc_htseq job')\n",
    "            elif gc == True:\n",
    "                if all_htseq == False or custom3_htseq == False or gc_htseq == False:\n",
    "                    if all_htseq == False:\n",
    "                        if 'j5a_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5a_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting all_htseq job')\n",
    "                    if custom3_htseq == False:\n",
    "                        if 'j5c_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5c_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting custom3_htseq job')\n",
    "                    if gc_htseq == False:\n",
    "                        if 'j5d_' + u_id in qstat_jid:\n",
    "                            print('')\n",
    "                            print('j5d_' + u_id + 'already running')\n",
    "                        else:\n",
    "                            print('')\n",
    "                            print('Submitting gc_htseq job')    \n",
    "\n",
    "\n",
    "### 3. Limit jobs to never take up more than 100 cores ###\n",
    "\n",
    "#elif slots > 100:\n",
    "#    print('Too many cores being used, removing all except first 6 jobs')\n",
    "#    qstat = pd.read_csv(script_dir + '/qstat2.txt', header = None)\n",
    "#    for line in qstat.iterrows():\n",
    "#        j_id = str.split(str(line[1]))[1]\n",
    "#        print(j_id)\n",
    "#        os.system('qdel ' + j_id)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class path in module path:\n",
      "\n",
      "class path(Path)\n",
      " |  Represents a filesystem path.\n",
      " |  \n",
      " |  For documentation on individual methods, consult their\n",
      " |  counterparts in :mod:`os.path`.\n",
      " |  \n",
      " |  Some methods are additionally included from :mod:`shutil`.\n",
      " |  The functions are linked directly into the class namespace\n",
      " |  such that they will be bound to the Path instance. For example,\n",
      " |  ``Path(src).copy(target)`` is equivalent to\n",
      " |  ``shutil.copy(src, target)``. Therefore, when referencing\n",
      " |  the docs for these methods, assume `src` references `self`,\n",
      " |  the Path instance.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      path\n",
      " |      Path\n",
      " |      builtins.str\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __new__(cls, *args, **kwargs)\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Path:\n",
      " |  \n",
      " |  __add__(self, more)\n",
      " |      Return self+value.\n",
      " |  \n",
      " |  __div__(self, rel)\n",
      " |      fp.__div__(rel) == fp / rel == fp.joinpath(rel)\n",
      " |      \n",
      " |      Join two path components, adding a separator character if\n",
      " |      needed.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.join`\n",
      " |  \n",
      " |  __enter__(self)\n",
      " |  \n",
      " |  __exit__(self, *_)\n",
      " |  \n",
      " |  __init__(self, other='')\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |  \n",
      " |  __rdiv__(self, rel)\n",
      " |      fp.__rdiv__(rel) == rel / fp\n",
      " |      \n",
      " |      Join two path components, adding a separator character if\n",
      " |      needed.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.join`\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __rtruediv__ = __rdiv__(self, rel)\n",
      " |      fp.__rdiv__(rel) == rel / fp\n",
      " |      \n",
      " |      Join two path components, adding a separator character if\n",
      " |      needed.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.join`\n",
      " |  \n",
      " |  __truediv__ = __div__(self, rel)\n",
      " |      fp.__div__(rel) == fp / rel == fp.joinpath(rel)\n",
      " |      \n",
      " |      Join two path components, adding a separator character if\n",
      " |      needed.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.join`\n",
      " |  \n",
      " |  abspath(self)\n",
      " |      .. seealso:: :func:`os.path.abspath`\n",
      " |  \n",
      " |  access(self, mode)\n",
      " |      Return ``True`` if current user has access to this path.\n",
      " |      \n",
      " |      mode - One of the constants :data:`os.F_OK`, :data:`os.R_OK`,\n",
      " |      :data:`os.W_OK`, :data:`os.X_OK`\n",
      " |      \n",
      " |      .. seealso:: :func:`os.access`\n",
      " |  \n",
      " |  basename(self)\n",
      " |      .. seealso:: :attr:`name`, :func:`os.path.basename`\n",
      " |  \n",
      " |  bytes(self)\n",
      " |      Open this file, read all bytes, return them as a string.\n",
      " |  \n",
      " |  cd = chdir(self)\n",
      " |      .. seealso:: :func:`os.chdir`\n",
      " |  \n",
      " |  chdir(self)\n",
      " |      .. seealso:: :func:`os.chdir`\n",
      " |  \n",
      " |  chmod(self, mode)\n",
      " |      Set the mode. May be the new mode (os.chmod behavior) or a `symbolic\n",
      " |      mode <http://en.wikipedia.org/wiki/Chmod#Symbolic_modes>`_.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.chmod`\n",
      " |  \n",
      " |  chown(self, uid=-1, gid=-1)\n",
      " |      Change the owner and group by names rather than the uid or gid numbers.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.chown`\n",
      " |  \n",
      " |  chroot(self)\n",
      " |      .. seealso:: :func:`os.chroot`\n",
      " |  \n",
      " |  chunks(self, size, *args, **kwargs)\n",
      " |      Returns a generator yielding chunks of the file, so it can\n",
      " |       be read piece by piece with a simple for loop.\n",
      " |      \n",
      " |      Any argument you pass after `size` will be passed to :meth:`open`.\n",
      " |      \n",
      " |      :example:\n",
      " |      \n",
      " |          >>> hash = hashlib.md5()\n",
      " |          >>> for chunk in Path(\"path.py\").chunks(8192, mode='rb'):\n",
      " |          ...     hash.update(chunk)\n",
      " |      \n",
      " |       This will read the file by chunks of 8192 bytes.\n",
      " |  \n",
      " |  copy(src, dst, *, follow_symlinks=True)\n",
      " |      Copy data and mode bits (\"cp src dst\"). Return the file's destination.\n",
      " |      \n",
      " |      The destination may be a directory.\n",
      " |      \n",
      " |      If follow_symlinks is false, symlinks won't be followed. This\n",
      " |      resembles GNU's \"cp -P src dst\".\n",
      " |      \n",
      " |      If source and destination are the same file, a SameFileError will be\n",
      " |      raised.\n",
      " |  \n",
      " |  copy2(src, dst, *, follow_symlinks=True)\n",
      " |      Copy data and all stat info (\"cp -p src dst\"). Return the file's\n",
      " |      destination.\"\n",
      " |      \n",
      " |      The destination may be a directory.\n",
      " |      \n",
      " |      If follow_symlinks is false, symlinks won't be followed. This\n",
      " |      resembles GNU's \"cp -P src dst\".\n",
      " |  \n",
      " |  copyfile(src, dst, *, follow_symlinks=True)\n",
      " |      Copy data from src to dst.\n",
      " |      \n",
      " |      If follow_symlinks is not set and src is a symbolic link, a new\n",
      " |      symlink will be created instead of copying the file it points to.\n",
      " |  \n",
      " |  copymode(src, dst, *, follow_symlinks=True)\n",
      " |      Copy mode bits from src to dst.\n",
      " |      \n",
      " |      If follow_symlinks is not set, symlinks aren't followed if and only\n",
      " |      if both `src` and `dst` are symlinks.  If `lchmod` isn't available\n",
      " |      (e.g. Linux) this method does nothing.\n",
      " |  \n",
      " |  copystat(src, dst, *, follow_symlinks=True)\n",
      " |      Copy all stat info (mode bits, atime, mtime, flags) from src to dst.\n",
      " |      \n",
      " |      If the optional flag `follow_symlinks` is not set, symlinks aren't followed if and\n",
      " |      only if both `src` and `dst` are symlinks.\n",
      " |  \n",
      " |  copytree(src, dst, symlinks=False, ignore=None, copy_function=<function copy2 at 0x102b9f0d0>, ignore_dangling_symlinks=False)\n",
      " |      Recursively copy a directory tree.\n",
      " |      \n",
      " |      The destination directory must not already exist.\n",
      " |      If exception(s) occur, an Error is raised with a list of reasons.\n",
      " |      \n",
      " |      If the optional symlinks flag is true, symbolic links in the\n",
      " |      source tree result in symbolic links in the destination tree; if\n",
      " |      it is false, the contents of the files pointed to by symbolic\n",
      " |      links are copied. If the file pointed by the symlink doesn't\n",
      " |      exist, an exception will be added in the list of errors raised in\n",
      " |      an Error exception at the end of the copy process.\n",
      " |      \n",
      " |      You can set the optional ignore_dangling_symlinks flag to true if you\n",
      " |      want to silence this exception. Notice that this has no effect on\n",
      " |      platforms that don't support os.symlink.\n",
      " |      \n",
      " |      The optional ignore argument is a callable. If given, it\n",
      " |      is called with the `src` parameter, which is the directory\n",
      " |      being visited by copytree(), and `names` which is the list of\n",
      " |      `src` contents, as returned by os.listdir():\n",
      " |      \n",
      " |          callable(src, names) -> ignored_names\n",
      " |      \n",
      " |      Since copytree() is called recursively, the callable will be\n",
      " |      called once for each directory that is copied. It returns a\n",
      " |      list of names relative to the `src` directory that should\n",
      " |      not be copied.\n",
      " |      \n",
      " |      The optional copy_function argument is a callable that will be used\n",
      " |      to copy each file. It will be called with the source path and the\n",
      " |      destination path as arguments. By default, copy2() is used, but any\n",
      " |      function that supports the same signature (like copy()) can be used.\n",
      " |  \n",
      " |  dirname(self)\n",
      " |      .. seealso:: :attr:`parent`, :func:`os.path.dirname`\n",
      " |  \n",
      " |  dirs(self, pattern=None)\n",
      " |      D.dirs() -> List of this directory's subdirectories.\n",
      " |      \n",
      " |      The elements of the list are Path objects.\n",
      " |      This does not walk recursively into subdirectories\n",
      " |      (but see :meth:`walkdirs`).\n",
      " |      \n",
      " |      With the optional `pattern` argument, this only lists\n",
      " |      directories whose names match the given pattern.  For\n",
      " |      example, ``d.dirs('build-*')``.\n",
      " |  \n",
      " |  exists(self)\n",
      " |      .. seealso:: :func:`os.path.exists`\n",
      " |  \n",
      " |  expand(self)\n",
      " |      Clean up a filename by calling :meth:`expandvars()`,\n",
      " |      :meth:`expanduser()`, and :meth:`normpath()` on it.\n",
      " |      \n",
      " |      This is commonly everything needed to clean up a filename\n",
      " |      read from a configuration file, for example.\n",
      " |  \n",
      " |  expanduser(self)\n",
      " |      .. seealso:: :func:`os.path.expanduser`\n",
      " |  \n",
      " |  expandvars(self)\n",
      " |      .. seealso:: :func:`os.path.expandvars`\n",
      " |  \n",
      " |  files(self, pattern=None)\n",
      " |      D.files() -> List of the files in this directory.\n",
      " |      \n",
      " |      The elements of the list are Path objects.\n",
      " |      This does not walk into subdirectories (see :meth:`walkfiles`).\n",
      " |      \n",
      " |      With the optional `pattern` argument, this only lists files\n",
      " |      whose names match the given pattern.  For example,\n",
      " |      ``d.files('*.pyc')``.\n",
      " |  \n",
      " |  fnmatch(self, pattern, normcase=None)\n",
      " |      Return ``True`` if `self.name` matches the given `pattern`.\n",
      " |      \n",
      " |      `pattern` - A filename pattern with wildcards,\n",
      " |          for example ``'*.py'``. If the pattern contains a `normcase`\n",
      " |          attribute, it is applied to the name and path prior to comparison.\n",
      " |      \n",
      " |      `normcase` - (optional) A function used to normalize the pattern and\n",
      " |          filename before matching. Defaults to :meth:`self.module`, which defaults\n",
      " |          to :meth:`os.path.normcase`.\n",
      " |      \n",
      " |      .. seealso:: :func:`fnmatch.fnmatch`\n",
      " |  \n",
      " |  get_owner = __get_owner_unix(self)\n",
      " |      Return the name of the owner of this file or directory. Follow\n",
      " |      symbolic links.\n",
      " |      \n",
      " |      .. seealso:: :attr:`owner`\n",
      " |  \n",
      " |  getatime(self)\n",
      " |      .. seealso:: :attr:`atime`, :func:`os.path.getatime`\n",
      " |  \n",
      " |  getctime(self)\n",
      " |      .. seealso:: :attr:`ctime`, :func:`os.path.getctime`\n",
      " |  \n",
      " |  getmtime(self)\n",
      " |      .. seealso:: :attr:`mtime`, :func:`os.path.getmtime`\n",
      " |  \n",
      " |  getsize(self)\n",
      " |      .. seealso:: :attr:`size`, :func:`os.path.getsize`\n",
      " |  \n",
      " |  glob(self, pattern)\n",
      " |      Return a list of Path objects that match the pattern.\n",
      " |      \n",
      " |      `pattern` - a path relative to this directory, with wildcards.\n",
      " |      \n",
      " |      For example, ``Path('/users').glob('*/bin/*')`` returns a list\n",
      " |      of all the files users have in their :file:`bin` directories.\n",
      " |      \n",
      " |      .. seealso:: :func:`glob.glob`\n",
      " |  \n",
      " |  in_place(self, mode='r', buffering=-1, encoding=None, errors=None, newline=None, backup_extension=None)\n",
      " |      A context in which a file may be re-written in-place with new content.\n",
      " |      \n",
      " |      Yields a tuple of :samp:`({readable}, {writable})` file objects, where `writable`\n",
      " |      replaces `readable`.\n",
      " |      \n",
      " |      If an exception occurs, the old file is restored, removing the\n",
      " |      written data.\n",
      " |      \n",
      " |      Mode *must not* use ``'w'``, ``'a'``, or ``'+'``; only read-only-modes are\n",
      " |      allowed. A :exc:`ValueError` is raised on invalid modes.\n",
      " |      \n",
      " |      For example, to add line numbers to a file::\n",
      " |      \n",
      " |          p = Path(filename)\n",
      " |          assert p.isfile()\n",
      " |          with p.in_place() as (reader, writer):\n",
      " |              for number, line in enumerate(reader, 1):\n",
      " |                  writer.write('{0:3}: '.format(number)))\n",
      " |                  writer.write(line)\n",
      " |      \n",
      " |      Thereafter, the file at `filename` will have line numbers in it.\n",
      " |  \n",
      " |  isabs(self)\n",
      " |      .. seealso:: :func:`os.path.isabs`\n",
      " |  \n",
      " |  isdir(self)\n",
      " |      .. seealso:: :func:`os.path.isdir`\n",
      " |  \n",
      " |  isfile(self)\n",
      " |      .. seealso:: :func:`os.path.isfile`\n",
      " |  \n",
      " |  islink(self)\n",
      " |      .. seealso:: :func:`os.path.islink`\n",
      " |  \n",
      " |  ismount(self)\n",
      " |      .. seealso:: :func:`os.path.ismount`\n",
      " |  \n",
      " |  lines(self, encoding=None, errors='strict', retain=True)\n",
      " |      Open this file, read all lines, return them in a list.\n",
      " |      \n",
      " |      Optional arguments:\n",
      " |          `encoding` - The Unicode encoding (or character set) of\n",
      " |              the file.  The default is ``None``, meaning the content\n",
      " |              of the file is read as 8-bit characters and returned\n",
      " |              as a list of (non-Unicode) str objects.\n",
      " |          `errors` - How to handle Unicode errors; see help(str.decode)\n",
      " |              for the options.  Default is ``'strict'``.\n",
      " |          `retain` - If ``True``, retain newline characters; but all newline\n",
      " |              character combinations (``'\\r'``, ``'\\n'``, ``'\\r\\n'``) are\n",
      " |              translated to ``'\\n'``.  If ``False``, newline characters are\n",
      " |              stripped off.  Default is ``True``.\n",
      " |      \n",
      " |      This uses ``'U'`` mode.\n",
      " |      \n",
      " |      .. seealso:: :meth:`text`\n",
      " |  \n",
      " |  link(self, newpath)\n",
      " |      Create a hard link at `newpath`, pointing to this file.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.link`\n",
      " |  \n",
      " |  listdir(self, pattern=None)\n",
      " |      D.listdir() -> List of items in this directory.\n",
      " |      \n",
      " |      Use :meth:`files` or :meth:`dirs` instead if you want a listing\n",
      " |      of just files or just subdirectories.\n",
      " |      \n",
      " |      The elements of the list are Path objects.\n",
      " |      \n",
      " |      With the optional `pattern` argument, this only lists\n",
      " |      items whose names match the given pattern.\n",
      " |      \n",
      " |      .. seealso:: :meth:`files`, :meth:`dirs`\n",
      " |  \n",
      " |  lstat(self)\n",
      " |      Like :meth:`stat`, but do not follow symbolic links.\n",
      " |      \n",
      " |      .. seealso:: :meth:`stat`, :func:`os.lstat`\n",
      " |  \n",
      " |  makedirs(self, mode=511)\n",
      " |      .. seealso:: :func:`os.makedirs`\n",
      " |  \n",
      " |  makedirs_p(self, mode=511)\n",
      " |      Like :meth:`makedirs`, but does not raise an exception if the\n",
      " |      directory already exists.\n",
      " |  \n",
      " |  merge_tree(self, dst, symlinks=False, *args, **kwargs)\n",
      " |      Copy entire contents of self to dst, overwriting existing\n",
      " |      contents in dst with those in self.\n",
      " |      \n",
      " |      If the additional keyword `update` is True, each\n",
      " |      `src` will only be copied if `dst` does not exist,\n",
      " |      or `src` is newer than `dst`.\n",
      " |      \n",
      " |      Note that the technique employed stages the files in a temporary\n",
      " |      directory first, so this function is not suitable for merging\n",
      " |      trees with large files, especially if the temporary directory\n",
      " |      is not capable of storing a copy of the entire source tree.\n",
      " |  \n",
      " |  mkdir(self, mode=511)\n",
      " |      .. seealso:: :func:`os.mkdir`\n",
      " |  \n",
      " |  mkdir_p(self, mode=511)\n",
      " |      Like :meth:`mkdir`, but does not raise an exception if the\n",
      " |      directory already exists.\n",
      " |  \n",
      " |  move(src, dst, copy_function=<function copy2 at 0x102b9f0d0>)\n",
      " |      Recursively move a file or directory to another location. This is\n",
      " |      similar to the Unix \"mv\" command. Return the file or directory's\n",
      " |      destination.\n",
      " |      \n",
      " |      If the destination is a directory or a symlink to a directory, the source\n",
      " |      is moved inside the directory. The destination path must not already\n",
      " |      exist.\n",
      " |      \n",
      " |      If the destination already exists but is not a directory, it may be\n",
      " |      overwritten depending on os.rename() semantics.\n",
      " |      \n",
      " |      If the destination is on our current filesystem, then rename() is used.\n",
      " |      Otherwise, src is copied to the destination and then removed. Symlinks are\n",
      " |      recreated under the new name if os.rename() fails because of cross\n",
      " |      filesystem renames.\n",
      " |      \n",
      " |      The optional `copy_function` argument is a callable that will be used\n",
      " |      to copy the source or it will be delegated to `copytree`.\n",
      " |      By default, copy2() is used, but any function that supports the same\n",
      " |      signature (like copy()) can be used.\n",
      " |      \n",
      " |      A lot more could be done here...  A look at a mv.c shows a lot of\n",
      " |      the issues this implementation glosses over.\n",
      " |  \n",
      " |  normcase(self)\n",
      " |      .. seealso:: :func:`os.path.normcase`\n",
      " |  \n",
      " |  normpath(self)\n",
      " |      .. seealso:: :func:`os.path.normpath`\n",
      " |  \n",
      " |  open(self, *args, **kwargs)\n",
      " |      Open this file and return a corresponding :class:`file` object.\n",
      " |      \n",
      " |      Keyword arguments work as in :func:`io.open`.  If the file cannot be\n",
      " |      opened, an :class:`~exceptions.OSError` is raised.\n",
      " |  \n",
      " |  pathconf(self, name)\n",
      " |      .. seealso:: :func:`os.pathconf`\n",
      " |  \n",
      " |  read_hash(self, hash_name)\n",
      " |      Calculate given hash for this file.\n",
      " |      \n",
      " |      List of supported hashes can be obtained from :mod:`hashlib` package.\n",
      " |      This reads the entire file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`hashlib.hash.digest`\n",
      " |  \n",
      " |  read_hexhash(self, hash_name)\n",
      " |      Calculate given hash for this file, returning hexdigest.\n",
      " |      \n",
      " |      List of supported hashes can be obtained from :mod:`hashlib` package.\n",
      " |      This reads the entire file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`hashlib.hash.hexdigest`\n",
      " |  \n",
      " |  read_md5(self)\n",
      " |      Calculate the md5 hash for this file.\n",
      " |      \n",
      " |      This reads through the entire file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`read_hash`\n",
      " |  \n",
      " |  readlink(self)\n",
      " |      Return the path to which this symbolic link points.\n",
      " |      \n",
      " |      The result may be an absolute or a relative path.\n",
      " |      \n",
      " |      .. seealso:: :meth:`readlinkabs`, :func:`os.readlink`\n",
      " |  \n",
      " |  readlinkabs(self)\n",
      " |      Return the path to which this symbolic link points.\n",
      " |      \n",
      " |      The result is always an absolute path.\n",
      " |      \n",
      " |      .. seealso:: :meth:`readlink`, :func:`os.readlink`\n",
      " |  \n",
      " |  realpath(self)\n",
      " |      .. seealso:: :func:`os.path.realpath`\n",
      " |  \n",
      " |  relpath(self, start='.')\n",
      " |      Return this path as a relative path,\n",
      " |      based from `start`, which defaults to the current working directory.\n",
      " |  \n",
      " |  relpathto(self, dest)\n",
      " |      Return a relative path from `self` to `dest`.\n",
      " |      \n",
      " |      If there is no relative path from `self` to `dest`, for example if\n",
      " |      they reside on different drives in Windows, then this returns\n",
      " |      ``dest.abspath()``.\n",
      " |  \n",
      " |  remove(self)\n",
      " |      .. seealso:: :func:`os.remove`\n",
      " |  \n",
      " |  remove_p(self)\n",
      " |      Like :meth:`remove`, but does not raise an exception if the\n",
      " |      file does not exist.\n",
      " |  \n",
      " |  removedirs(self)\n",
      " |      .. seealso:: :func:`os.removedirs`\n",
      " |  \n",
      " |  removedirs_p(self)\n",
      " |      Like :meth:`removedirs`, but does not raise an exception if the\n",
      " |      directory is not empty or does not exist.\n",
      " |  \n",
      " |  rename(self, new)\n",
      " |      .. seealso:: :func:`os.rename`\n",
      " |  \n",
      " |  renames(self, new)\n",
      " |      .. seealso:: :func:`os.renames`\n",
      " |  \n",
      " |  rmdir(self)\n",
      " |      .. seealso:: :func:`os.rmdir`\n",
      " |  \n",
      " |  rmdir_p(self)\n",
      " |      Like :meth:`rmdir`, but does not raise an exception if the\n",
      " |      directory is not empty or does not exist.\n",
      " |  \n",
      " |  rmtree(path, ignore_errors=False, onerror=None)\n",
      " |      Recursively delete a directory tree.\n",
      " |      \n",
      " |      If ignore_errors is set, errors are ignored; otherwise, if onerror\n",
      " |      is set, it is called to handle the error with arguments (func,\n",
      " |      path, exc_info) where func is platform and implementation dependent;\n",
      " |      path is the argument to that function that caused it to fail; and\n",
      " |      exc_info is a tuple returned by sys.exc_info().  If ignore_errors\n",
      " |      is false and onerror is None, an exception is raised.\n",
      " |  \n",
      " |  rmtree_p(self)\n",
      " |      Like :meth:`rmtree`, but does not raise an exception if the\n",
      " |      directory does not exist.\n",
      " |  \n",
      " |  samefile(self, other)\n",
      " |      .. seealso:: :func:`os.path.samefile`\n",
      " |  \n",
      " |  splitall(self)\n",
      " |      Return a list of the path components in this path.\n",
      " |      \n",
      " |      The first item in the list will be a Path.  Its value will be\n",
      " |      either :data:`os.curdir`, :data:`os.pardir`, empty, or the root\n",
      " |      directory of this path (for example, ``'/'`` or ``'C:\\\\'``).  The\n",
      " |      other items in the list will be strings.\n",
      " |      \n",
      " |      ``path.Path.joinpath(*result)`` will yield the original path.\n",
      " |  \n",
      " |  splitdrive(self)\n",
      " |      p.splitdrive() -> Return ``(p.drive, <the rest of p>)``.\n",
      " |      \n",
      " |      Split the drive specifier from this path.  If there is\n",
      " |      no drive specifier, :samp:`{p.drive}` is empty, so the return value\n",
      " |      is simply ``(Path(''), p)``.  This is always the case on Unix.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.splitdrive`\n",
      " |  \n",
      " |  splitext(self)\n",
      " |      p.splitext() -> Return ``(p.stripext(), p.ext)``.\n",
      " |      \n",
      " |      Split the filename extension from this path and return\n",
      " |      the two parts.  Either part may be empty.\n",
      " |      \n",
      " |      The extension is everything from ``'.'`` to the end of the\n",
      " |      last path segment.  This has the property that if\n",
      " |      ``(a, b) == p.splitext()``, then ``a + b == p``.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.path.splitext`\n",
      " |  \n",
      " |  splitpath(self)\n",
      " |      p.splitpath() -> Return ``(p.parent, p.name)``.\n",
      " |      \n",
      " |      .. seealso:: :attr:`parent`, :attr:`name`, :func:`os.path.split`\n",
      " |  \n",
      " |  splitunc(self)\n",
      " |      .. seealso:: :func:`os.path.splitunc`\n",
      " |  \n",
      " |  stat(self)\n",
      " |      Perform a ``stat()`` system call on this path.\n",
      " |      \n",
      " |      .. seealso:: :meth:`lstat`, :func:`os.stat`\n",
      " |  \n",
      " |  statvfs(self)\n",
      " |      Perform a ``statvfs()`` system call on this path.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.statvfs`\n",
      " |  \n",
      " |  stripext(self)\n",
      " |      p.stripext() -> Remove one file extension from the path.\n",
      " |      \n",
      " |      For example, ``Path('/home/guido/python.tar.gz').stripext()``\n",
      " |      returns ``Path('/home/guido/python.tar')``.\n",
      " |  \n",
      " |  symlink(self, newlink)\n",
      " |      Create a symbolic link at `newlink`, pointing here.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.symlink`\n",
      " |  \n",
      " |  text(self, encoding=None, errors='strict')\n",
      " |      Open this file, read it in, return the content as a string.\n",
      " |      \n",
      " |      All newline sequences are converted to ``'\\n'``.  Keyword arguments\n",
      " |      will be passed to :meth:`open`.\n",
      " |      \n",
      " |      .. seealso:: :meth:`lines`\n",
      " |  \n",
      " |  touch(self)\n",
      " |      Set the access/modified times of this file to the current time.\n",
      " |      Create the file if it does not exist.\n",
      " |  \n",
      " |  unlink(self)\n",
      " |      .. seealso:: :func:`os.unlink`\n",
      " |  \n",
      " |  unlink_p(self)\n",
      " |      Like :meth:`unlink`, but does not raise an exception if the\n",
      " |      file does not exist.\n",
      " |  \n",
      " |  utime(self, times)\n",
      " |      Set the access and modified times of this file.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.utime`\n",
      " |  \n",
      " |  walk(self, pattern=None, errors='strict')\n",
      " |      D.walk() -> iterator over files and subdirs, recursively.\n",
      " |      \n",
      " |      The iterator yields Path objects naming each child item of\n",
      " |      this directory and its descendants.  This requires that\n",
      " |      ``D.isdir()``.\n",
      " |      \n",
      " |      This performs a depth-first traversal of the directory tree.\n",
      " |      Each directory is returned just before all its children.\n",
      " |      \n",
      " |      The `errors=` keyword argument controls behavior when an\n",
      " |      error occurs.  The default is ``'strict'``, which causes an\n",
      " |      exception.  Other allowed values are ``'warn'`` (which\n",
      " |      reports the error via :func:`warnings.warn()`), and ``'ignore'``.\n",
      " |      `errors` may also be an arbitrary callable taking a msg parameter.\n",
      " |  \n",
      " |  walkdirs(self, pattern=None, errors='strict')\n",
      " |      D.walkdirs() -> iterator over subdirs, recursively.\n",
      " |      \n",
      " |      With the optional `pattern` argument, this yields only\n",
      " |      directories whose names match the given pattern.  For\n",
      " |      example, ``mydir.walkdirs('*test')`` yields only directories\n",
      " |      with names ending in ``'test'``.\n",
      " |      \n",
      " |      The `errors=` keyword argument controls behavior when an\n",
      " |      error occurs.  The default is ``'strict'``, which causes an\n",
      " |      exception.  The other allowed values are ``'warn'`` (which\n",
      " |      reports the error via :func:`warnings.warn()`), and ``'ignore'``.\n",
      " |  \n",
      " |  walkfiles(self, pattern=None, errors='strict')\n",
      " |      D.walkfiles() -> iterator over files in D, recursively.\n",
      " |      \n",
      " |      The optional argument `pattern` limits the results to files\n",
      " |      with names that match the pattern.  For example,\n",
      " |      ``mydir.walkfiles('*.tmp')`` yields only files with the ``.tmp``\n",
      " |      extension.\n",
      " |  \n",
      " |  write_bytes(self, bytes, append=False)\n",
      " |      Open this file and write the given bytes to it.\n",
      " |      \n",
      " |      Default behavior is to overwrite any existing file.\n",
      " |      Call ``p.write_bytes(bytes, append=True)`` to append instead.\n",
      " |  \n",
      " |  write_lines(self, lines, encoding=None, errors='strict', linesep='\\n', append=False)\n",
      " |      Write the given lines of text to this file.\n",
      " |      \n",
      " |      By default this overwrites any existing file at this path.\n",
      " |      \n",
      " |      This puts a platform-specific newline sequence on every line.\n",
      " |      See `linesep` below.\n",
      " |      \n",
      " |          `lines` - A list of strings.\n",
      " |      \n",
      " |          `encoding` - A Unicode encoding to use.  This applies only if\n",
      " |              `lines` contains any Unicode strings.\n",
      " |      \n",
      " |          `errors` - How to handle errors in Unicode encoding.  This\n",
      " |              also applies only to Unicode strings.\n",
      " |      \n",
      " |          linesep - The desired line-ending.  This line-ending is\n",
      " |              applied to every line.  If a line already has any\n",
      " |              standard line ending (``'\\r'``, ``'\\n'``, ``'\\r\\n'``,\n",
      " |              ``u'\\x85'``, ``u'\\r\\x85'``, ``u'\\u2028'``), that will\n",
      " |              be stripped off and this will be used instead.  The\n",
      " |              default is os.linesep, which is platform-dependent\n",
      " |              (``'\\r\\n'`` on Windows, ``'\\n'`` on Unix, etc.).\n",
      " |              Specify ``None`` to write the lines as-is, like\n",
      " |              :meth:`file.writelines`.\n",
      " |      \n",
      " |      Use the keyword argument ``append=True`` to append lines to the\n",
      " |      file.  The default is to overwrite the file.\n",
      " |      \n",
      " |      .. warning ::\n",
      " |      \n",
      " |          When you use this with Unicode data, if the encoding of the\n",
      " |          existing data in the file is different from the encoding\n",
      " |          you specify with the `encoding=` parameter, the result is\n",
      " |          mixed-encoding data, which can really confuse someone trying\n",
      " |          to read the file later.\n",
      " |  \n",
      " |  write_text(self, text, encoding=None, errors='strict', linesep='\\n', append=False)\n",
      " |      Write the given text to this file.\n",
      " |      \n",
      " |      The default behavior is to overwrite any existing file;\n",
      " |      to append instead, use the `append=True` keyword argument.\n",
      " |      \n",
      " |      There are two differences between :meth:`write_text` and\n",
      " |      :meth:`write_bytes`: newline handling and Unicode handling.\n",
      " |      See below.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      \n",
      " |        `text` - str/unicode - The text to be written.\n",
      " |      \n",
      " |        `encoding` - str - The Unicode encoding that will be used.\n",
      " |            This is ignored if `text` isn't a Unicode string.\n",
      " |      \n",
      " |        `errors` - str - How to handle Unicode encoding errors.\n",
      " |            Default is ``'strict'``.  See ``help(unicode.encode)`` for the\n",
      " |            options.  This is ignored if `text` isn't a Unicode\n",
      " |            string.\n",
      " |      \n",
      " |        `linesep` - keyword argument - str/unicode - The sequence of\n",
      " |            characters to be used to mark end-of-line.  The default is\n",
      " |            :data:`os.linesep`.  You can also specify ``None`` to\n",
      " |            leave all newlines as they are in `text`.\n",
      " |      \n",
      " |        `append` - keyword argument - bool - Specifies what to do if\n",
      " |            the file already exists (``True``: append to the end of it;\n",
      " |            ``False``: overwrite it.)  The default is ``False``.\n",
      " |      \n",
      " |      \n",
      " |      --- Newline handling.\n",
      " |      \n",
      " |      ``write_text()`` converts all standard end-of-line sequences\n",
      " |      (``'\\n'``, ``'\\r'``, and ``'\\r\\n'``) to your platform's default\n",
      " |      end-of-line sequence (see :data:`os.linesep`; on Windows, for example,\n",
      " |      the end-of-line marker is ``'\\r\\n'``).\n",
      " |      \n",
      " |      If you don't like your platform's default, you can override it\n",
      " |      using the `linesep=` keyword argument.  If you specifically want\n",
      " |      ``write_text()`` to preserve the newlines as-is, use ``linesep=None``.\n",
      " |      \n",
      " |      This applies to Unicode text the same as to 8-bit text, except\n",
      " |      there are three additional standard Unicode end-of-line sequences:\n",
      " |      ``u'\\x85'``, ``u'\\r\\x85'``, and ``u'\\u2028'``.\n",
      " |      \n",
      " |      (This is slightly different from when you open a file for\n",
      " |      writing with ``fopen(filename, \"w\")`` in C or ``open(filename, 'w')``\n",
      " |      in Python.)\n",
      " |      \n",
      " |      \n",
      " |      --- Unicode\n",
      " |      \n",
      " |      If `text` isn't Unicode, then apart from newline handling, the\n",
      " |      bytes are written verbatim to the file.  The `encoding` and\n",
      " |      `errors` arguments are not used and must be omitted.\n",
      " |      \n",
      " |      If `text` is Unicode, it is first converted to :func:`bytes` using the\n",
      " |      specified `encoding` (or the default encoding if `encoding`\n",
      " |      isn't specified).  The `errors` argument applies only to this\n",
      " |      conversion.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from Path:\n",
      " |  \n",
      " |  getcwd() from builtins.type\n",
      " |      Return the current working directory as a path object.\n",
      " |      \n",
      " |      .. seealso:: :func:`os.getcwdu`\n",
      " |  \n",
      " |  using_module = wrapper(module) from builtins.type\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Path:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  atime\n",
      " |      Last access time of the file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`getatime`, :func:`os.path.getatime`\n",
      " |  \n",
      " |  ctime\n",
      " |      Creation time of the file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`getctime`, :func:`os.path.getctime`\n",
      " |  \n",
      " |  drive\n",
      " |      The drive specifier, for example ``'C:'``.\n",
      " |      \n",
      " |      This is always empty on systems that don't use drive specifiers.\n",
      " |  \n",
      " |  ext\n",
      " |      The file extension, for example ``'.py'``.\n",
      " |  \n",
      " |  mtime\n",
      " |      Last-modified time of the file.\n",
      " |      \n",
      " |      .. seealso:: :meth:`getmtime`, :func:`os.path.getmtime`\n",
      " |  \n",
      " |  name\n",
      " |      The name of this file or directory without the full path.\n",
      " |      \n",
      " |      For example,\n",
      " |      ``Path('/usr/local/lib/libpython.so').name == 'libpython.so'``\n",
      " |      \n",
      " |      .. seealso:: :meth:`basename`, :func:`os.path.basename`\n",
      " |  \n",
      " |  namebase\n",
      " |      The same as :meth:`name`, but with one file extension stripped off.\n",
      " |      \n",
      " |      For example,\n",
      " |      ``Path('/home/guido/python.tar.gz').name == 'python.tar.gz'``,\n",
      " |      but\n",
      " |      ``Path('/home/guido/python.tar.gz').namebase == 'python.tar'``.\n",
      " |  \n",
      " |  owner\n",
      " |      Name of the owner of this file or directory.\n",
      " |      \n",
      " |      .. seealso:: :meth:`get_owner`\n",
      " |  \n",
      " |  parent\n",
      " |      This path's parent directory, as a new Path object.\n",
      " |      \n",
      " |      For example,\n",
      " |      ``Path('/usr/local/lib/libpython.so').parent ==\n",
      " |      Path('/usr/local/lib')``\n",
      " |      \n",
      " |      .. seealso:: :meth:`dirname`, :func:`os.path.dirname`\n",
      " |  \n",
      " |  size\n",
      " |      Size of the file, in bytes.\n",
      " |      \n",
      " |      .. seealso:: :meth:`getsize`, :func:`os.path.getsize`\n",
      " |  \n",
      " |  special\n",
      " |      classmethod(function) -> method\n",
      " |      \n",
      " |      Convert a function to be a class method.\n",
      " |      \n",
      " |      A class method receives the class as implicit first argument,\n",
      " |      just like an instance method receives the instance.\n",
      " |      To declare a class method, use this idiom:\n",
      " |      \n",
      " |        class C:\n",
      " |            def f(cls, arg1, arg2, ...): ...\n",
      " |            f = classmethod(f)\n",
      " |      \n",
      " |      It can be called either on the class (e.g. C.f()) or on an instance\n",
      " |      (e.g. C().f()).  The instance is ignored except for its class.\n",
      " |      If a class method is called for a derived class, the derived class\n",
      " |      object is passed as the implied first argument.\n",
      " |      \n",
      " |      Class methods are different than C++ or Java static methods.\n",
      " |      If you want those, see the staticmethod builtin.\n",
      " |  \n",
      " |  uncshare\n",
      " |      The UNC mount point for this path.\n",
      " |      This is empty for paths on local drives.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Path:\n",
      " |  \n",
      " |  joinpath = functools.partial(<function Path.joinpath at 0x1048eb620>, ...\n",
      " |      partial(func, *args, **keywords) - new function with partial application\n",
      " |      of the given arguments and keywords.\n",
      " |  \n",
      " |  module = <module 'posixpath' from '/Users/jamestorpy/anaconda3/lib/pyt...\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from builtins.str:\n",
      " |  \n",
      " |  __contains__(self, key, /)\n",
      " |      Return key in self.\n",
      " |  \n",
      " |  __eq__(self, value, /)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __format__(...)\n",
      " |      S.__format__(format_spec) -> str\n",
      " |      \n",
      " |      Return a formatted version of S as described by format_spec.\n",
      " |  \n",
      " |  __ge__(self, value, /)\n",
      " |      Return self>=value.\n",
      " |  \n",
      " |  __getattribute__(self, name, /)\n",
      " |      Return getattr(self, name).\n",
      " |  \n",
      " |  __getitem__(self, key, /)\n",
      " |      Return self[key].\n",
      " |  \n",
      " |  __getnewargs__(...)\n",
      " |  \n",
      " |  __gt__(self, value, /)\n",
      " |      Return self>value.\n",
      " |  \n",
      " |  __hash__(self, /)\n",
      " |      Return hash(self).\n",
      " |  \n",
      " |  __iter__(self, /)\n",
      " |      Implement iter(self).\n",
      " |  \n",
      " |  __le__(self, value, /)\n",
      " |      Return self<=value.\n",
      " |  \n",
      " |  __len__(self, /)\n",
      " |      Return len(self).\n",
      " |  \n",
      " |  __lt__(self, value, /)\n",
      " |      Return self<value.\n",
      " |  \n",
      " |  __mod__(self, value, /)\n",
      " |      Return self%value.\n",
      " |  \n",
      " |  __mul__(self, value, /)\n",
      " |      Return self*value.n\n",
      " |  \n",
      " |  __ne__(self, value, /)\n",
      " |      Return self!=value.\n",
      " |  \n",
      " |  __rmod__(self, value, /)\n",
      " |      Return value%self.\n",
      " |  \n",
      " |  __rmul__(self, value, /)\n",
      " |      Return self*value.\n",
      " |  \n",
      " |  __sizeof__(...)\n",
      " |      S.__sizeof__() -> size of S in memory, in bytes\n",
      " |  \n",
      " |  __str__(self, /)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  capitalize(...)\n",
      " |      S.capitalize() -> str\n",
      " |      \n",
      " |      Return a capitalized version of S, i.e. make the first character\n",
      " |      have upper case and the rest lower case.\n",
      " |  \n",
      " |  casefold(...)\n",
      " |      S.casefold() -> str\n",
      " |      \n",
      " |      Return a version of S suitable for caseless comparisons.\n",
      " |  \n",
      " |  center(...)\n",
      " |      S.center(width[, fillchar]) -> str\n",
      " |      \n",
      " |      Return S centered in a string of length width. Padding is\n",
      " |      done using the specified fill character (default is a space)\n",
      " |  \n",
      " |  count(...)\n",
      " |      S.count(sub[, start[, end]]) -> int\n",
      " |      \n",
      " |      Return the number of non-overlapping occurrences of substring sub in\n",
      " |      string S[start:end].  Optional arguments start and end are\n",
      " |      interpreted as in slice notation.\n",
      " |  \n",
      " |  encode(...)\n",
      " |      S.encode(encoding='utf-8', errors='strict') -> bytes\n",
      " |      \n",
      " |      Encode S using the codec registered for encoding. Default encoding\n",
      " |      is 'utf-8'. errors may be given to set a different error\n",
      " |      handling scheme. Default is 'strict' meaning that encoding errors raise\n",
      " |      a UnicodeEncodeError. Other possible values are 'ignore', 'replace' and\n",
      " |      'xmlcharrefreplace' as well as any other name registered with\n",
      " |      codecs.register_error that can handle UnicodeEncodeErrors.\n",
      " |  \n",
      " |  endswith(...)\n",
      " |      S.endswith(suffix[, start[, end]]) -> bool\n",
      " |      \n",
      " |      Return True if S ends with the specified suffix, False otherwise.\n",
      " |      With optional start, test S beginning at that position.\n",
      " |      With optional end, stop comparing S at that position.\n",
      " |      suffix can also be a tuple of strings to try.\n",
      " |  \n",
      " |  expandtabs(...)\n",
      " |      S.expandtabs(tabsize=8) -> str\n",
      " |      \n",
      " |      Return a copy of S where all tab characters are expanded using spaces.\n",
      " |      If tabsize is not given, a tab size of 8 characters is assumed.\n",
      " |  \n",
      " |  find(...)\n",
      " |      S.find(sub[, start[, end]]) -> int\n",
      " |      \n",
      " |      Return the lowest index in S where substring sub is found,\n",
      " |      such that sub is contained within S[start:end].  Optional\n",
      " |      arguments start and end are interpreted as in slice notation.\n",
      " |      \n",
      " |      Return -1 on failure.\n",
      " |  \n",
      " |  format(...)\n",
      " |      S.format(*args, **kwargs) -> str\n",
      " |      \n",
      " |      Return a formatted version of S, using substitutions from args and kwargs.\n",
      " |      The substitutions are identified by braces ('{' and '}').\n",
      " |  \n",
      " |  format_map(...)\n",
      " |      S.format_map(mapping) -> str\n",
      " |      \n",
      " |      Return a formatted version of S, using substitutions from mapping.\n",
      " |      The substitutions are identified by braces ('{' and '}').\n",
      " |  \n",
      " |  index(...)\n",
      " |      S.index(sub[, start[, end]]) -> int\n",
      " |      \n",
      " |      Like S.find() but raise ValueError when the substring is not found.\n",
      " |  \n",
      " |  isalnum(...)\n",
      " |      S.isalnum() -> bool\n",
      " |      \n",
      " |      Return True if all characters in S are alphanumeric\n",
      " |      and there is at least one character in S, False otherwise.\n",
      " |  \n",
      " |  isalpha(...)\n",
      " |      S.isalpha() -> bool\n",
      " |      \n",
      " |      Return True if all characters in S are alphabetic\n",
      " |      and there is at least one character in S, False otherwise.\n",
      " |  \n",
      " |  isdecimal(...)\n",
      " |      S.isdecimal() -> bool\n",
      " |      \n",
      " |      Return True if there are only decimal characters in S,\n",
      " |      False otherwise.\n",
      " |  \n",
      " |  isdigit(...)\n",
      " |      S.isdigit() -> bool\n",
      " |      \n",
      " |      Return True if all characters in S are digits\n",
      " |      and there is at least one character in S, False otherwise.\n",
      " |  \n",
      " |  isidentifier(...)\n",
      " |      S.isidentifier() -> bool\n",
      " |      \n",
      " |      Return True if S is a valid identifier according\n",
      " |      to the language definition.\n",
      " |      \n",
      " |      Use keyword.iskeyword() to test for reserved identifiers\n",
      " |      such as \"def\" and \"class\".\n",
      " |  \n",
      " |  islower(...)\n",
      " |      S.islower() -> bool\n",
      " |      \n",
      " |      Return True if all cased characters in S are lowercase and there is\n",
      " |      at least one cased character in S, False otherwise.\n",
      " |  \n",
      " |  isnumeric(...)\n",
      " |      S.isnumeric() -> bool\n",
      " |      \n",
      " |      Return True if there are only numeric characters in S,\n",
      " |      False otherwise.\n",
      " |  \n",
      " |  isprintable(...)\n",
      " |      S.isprintable() -> bool\n",
      " |      \n",
      " |      Return True if all characters in S are considered\n",
      " |      printable in repr() or S is empty, False otherwise.\n",
      " |  \n",
      " |  isspace(...)\n",
      " |      S.isspace() -> bool\n",
      " |      \n",
      " |      Return True if all characters in S are whitespace\n",
      " |      and there is at least one character in S, False otherwise.\n",
      " |  \n",
      " |  istitle(...)\n",
      " |      S.istitle() -> bool\n",
      " |      \n",
      " |      Return True if S is a titlecased string and there is at least one\n",
      " |      character in S, i.e. upper- and titlecase characters may only\n",
      " |      follow uncased characters and lowercase characters only cased ones.\n",
      " |      Return False otherwise.\n",
      " |  \n",
      " |  isupper(...)\n",
      " |      S.isupper() -> bool\n",
      " |      \n",
      " |      Return True if all cased characters in S are uppercase and there is\n",
      " |      at least one cased character in S, False otherwise.\n",
      " |  \n",
      " |  join(...)\n",
      " |      S.join(iterable) -> str\n",
      " |      \n",
      " |      Return a string which is the concatenation of the strings in the\n",
      " |      iterable.  The separator between elements is S.\n",
      " |  \n",
      " |  ljust(...)\n",
      " |      S.ljust(width[, fillchar]) -> str\n",
      " |      \n",
      " |      Return S left-justified in a Unicode string of length width. Padding is\n",
      " |      done using the specified fill character (default is a space).\n",
      " |  \n",
      " |  lower(...)\n",
      " |      S.lower() -> str\n",
      " |      \n",
      " |      Return a copy of the string S converted to lowercase.\n",
      " |  \n",
      " |  lstrip(...)\n",
      " |      S.lstrip([chars]) -> str\n",
      " |      \n",
      " |      Return a copy of the string S with leading whitespace removed.\n",
      " |      If chars is given and not None, remove characters in chars instead.\n",
      " |  \n",
      " |  partition(...)\n",
      " |      S.partition(sep) -> (head, sep, tail)\n",
      " |      \n",
      " |      Search for the separator sep in S, and return the part before it,\n",
      " |      the separator itself, and the part after it.  If the separator is not\n",
      " |      found, return S and two empty strings.\n",
      " |  \n",
      " |  replace(...)\n",
      " |      S.replace(old, new[, count]) -> str\n",
      " |      \n",
      " |      Return a copy of S with all occurrences of substring\n",
      " |      old replaced by new.  If the optional argument count is\n",
      " |      given, only the first count occurrences are replaced.\n",
      " |  \n",
      " |  rfind(...)\n",
      " |      S.rfind(sub[, start[, end]]) -> int\n",
      " |      \n",
      " |      Return the highest index in S where substring sub is found,\n",
      " |      such that sub is contained within S[start:end].  Optional\n",
      " |      arguments start and end are interpreted as in slice notation.\n",
      " |      \n",
      " |      Return -1 on failure.\n",
      " |  \n",
      " |  rindex(...)\n",
      " |      S.rindex(sub[, start[, end]]) -> int\n",
      " |      \n",
      " |      Like S.rfind() but raise ValueError when the substring is not found.\n",
      " |  \n",
      " |  rjust(...)\n",
      " |      S.rjust(width[, fillchar]) -> str\n",
      " |      \n",
      " |      Return S right-justified in a string of length width. Padding is\n",
      " |      done using the specified fill character (default is a space).\n",
      " |  \n",
      " |  rpartition(...)\n",
      " |      S.rpartition(sep) -> (head, sep, tail)\n",
      " |      \n",
      " |      Search for the separator sep in S, starting at the end of S, and return\n",
      " |      the part before it, the separator itself, and the part after it.  If the\n",
      " |      separator is not found, return two empty strings and S.\n",
      " |  \n",
      " |  rsplit(...)\n",
      " |      S.rsplit(sep=None, maxsplit=-1) -> list of strings\n",
      " |      \n",
      " |      Return a list of the words in S, using sep as the\n",
      " |      delimiter string, starting at the end of the string and\n",
      " |      working to the front.  If maxsplit is given, at most maxsplit\n",
      " |      splits are done. If sep is not specified, any whitespace string\n",
      " |      is a separator.\n",
      " |  \n",
      " |  rstrip(...)\n",
      " |      S.rstrip([chars]) -> str\n",
      " |      \n",
      " |      Return a copy of the string S with trailing whitespace removed.\n",
      " |      If chars is given and not None, remove characters in chars instead.\n",
      " |  \n",
      " |  split(...)\n",
      " |      S.split(sep=None, maxsplit=-1) -> list of strings\n",
      " |      \n",
      " |      Return a list of the words in S, using sep as the\n",
      " |      delimiter string.  If maxsplit is given, at most maxsplit\n",
      " |      splits are done. If sep is not specified or is None, any\n",
      " |      whitespace string is a separator and empty strings are\n",
      " |      removed from the result.\n",
      " |  \n",
      " |  splitlines(...)\n",
      " |      S.splitlines([keepends]) -> list of strings\n",
      " |      \n",
      " |      Return a list of the lines in S, breaking at line boundaries.\n",
      " |      Line breaks are not included in the resulting list unless keepends\n",
      " |      is given and true.\n",
      " |  \n",
      " |  startswith(...)\n",
      " |      S.startswith(prefix[, start[, end]]) -> bool\n",
      " |      \n",
      " |      Return True if S starts with the specified prefix, False otherwise.\n",
      " |      With optional start, test S beginning at that position.\n",
      " |      With optional end, stop comparing S at that position.\n",
      " |      prefix can also be a tuple of strings to try.\n",
      " |  \n",
      " |  strip(...)\n",
      " |      S.strip([chars]) -> str\n",
      " |      \n",
      " |      Return a copy of the string S with leading and trailing\n",
      " |      whitespace removed.\n",
      " |      If chars is given and not None, remove characters in chars instead.\n",
      " |  \n",
      " |  swapcase(...)\n",
      " |      S.swapcase() -> str\n",
      " |      \n",
      " |      Return a copy of S with uppercase characters converted to lowercase\n",
      " |      and vice versa.\n",
      " |  \n",
      " |  title(...)\n",
      " |      S.title() -> str\n",
      " |      \n",
      " |      Return a titlecased version of S, i.e. words start with title case\n",
      " |      characters, all remaining cased characters have lower case.\n",
      " |  \n",
      " |  translate(...)\n",
      " |      S.translate(table) -> str\n",
      " |      \n",
      " |      Return a copy of the string S in which each character has been mapped\n",
      " |      through the given translation table. The table must implement\n",
      " |      lookup/indexing via __getitem__, for instance a dictionary or list,\n",
      " |      mapping Unicode ordinals to Unicode ordinals, strings, or None. If\n",
      " |      this operation raises LookupError, the character is left untouched.\n",
      " |      Characters mapped to None are deleted.\n",
      " |  \n",
      " |  upper(...)\n",
      " |      S.upper() -> str\n",
      " |      \n",
      " |      Return a copy of S converted to uppercase.\n",
      " |  \n",
      " |  zfill(...)\n",
      " |      S.zfill(width) -> str\n",
      " |      \n",
      " |      Pad a numeric string S with zeros on the left, to fill a field\n",
      " |      of the specified width. The string S is never truncated.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from builtins.str:\n",
      " |  \n",
      " |  maketrans(x, y=None, z=None, /)\n",
      " |      Return a translation table usable for str.translate().\n",
      " |      \n",
      " |      If there is only one argument, it must be a dictionary mapping Unicode\n",
      " |      ordinals (integers) or characters to Unicode ordinals, strings or None.\n",
      " |      Character keys will be then converted to ordinals.\n",
      " |      If there are two arguments, they must be strings of equal length, and\n",
      " |      in the resulting dictionary, each character in x will be mapped to the\n",
      " |      character at the same position in y. If there is a third argument, it\n",
      " |      must be a string, whose characters will be mapped to None in the result.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "qstat = pd.read_csv(record_dir + '/qstat.txt', header = None)\n",
    "qstat_jid = qstat[0].str.split(' ').str[3]\n",
    "#print(qstat_jid)\n",
    "\n",
    "\n",
    "for row in qstat_jid:\n",
    "    if 'fq_prPT11' in row:\n",
    "        print('yep')\n",
    "    \n",
    "#for r in qstat_jid.iterrows():\n",
    "#    print(r[1]), type(r[1])\n",
    "\n",
    "#for row in qstat_jid.iterrows():\n",
    "#    print(row)\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['fq_prPT11', 'fq_prPT9']\n",
      "oops\n"
     ]
    }
   ],
   "source": [
    "print(qstat_jid)\n",
    "if 'fq_prPT9' in qstat_jid:\n",
    "    print('oops')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>650406 0.00513 fq_prPT11  jamtor       r     ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>650407 0.00513 fq_prPT9   jamtor       r     ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0\n",
       "0   650406 0.00513 fq_prPT11  jamtor       r     ...\n",
       "1   650407 0.00513 fq_prPT9   jamtor       r     ..."
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qstat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qstat = pd.read_csv(record_dir + '/qstat.txt', header = None)\n",
    "qstat_jid = qstat[0].str.split(' ').str[3]\n",
    "temp = list(qstat_jid)\n",
    "'fq_prPT10' in temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
